fdbserver/storageserver.actor.cpp

```diff
@@ -1235,7 +1235,12 @@ ACTOR Future<Void> getValueQ(StorageServer* data, GetValueRequest req) {
 			                      "getValueQ.DoRead"); //.detail("TaskID", g_network->getCurrentTask());
 
 		state Optional<Value> v;
-		state Version version = wait(waitForVersion(data, req.version, req.spanContext));
+		// If the client specified the latest commit version (that mutated the shard(s) being served
+		// by this storage server) then return the value that corresponds to that version.
+		Version readVersion = req.ssLatestCommitVersions.hasVersion(data->tag)
```

Comment:
when would `ssLatestCommitVersions` not have `data->tag` ?

Reply:
> * when would `ssLatestCommitVersions` not have `data->tag` ?
I can think of the following scenarios: (a) if  the read is being done at an older version (the client won't propagate the latest commit version in that case, as it doesn't need to) (b) when the cluster/sequencer starts the version vector will be empty (and tags won't be added to the version vector until a transaction mutates them) (b) when a new storage server is added to the system? (I don't know if we allow new storage servers to be added to a running cluster but if we do then I don't think we will have entries for those until a transaction mutates them).

> * Consider making this `if` statement an inline function, as I see it used elsewhere.

Good suggestion. Done.

> * we now have a knob `ENABLE_VERSION_VECTOR`. We should probably not access things like the version vector when the knob is false.

Since we are working on a branch I am not sure we need to protect all the new code with the knob.  But maybe we should protect some new code with the knob (until we are ready to test this feature end-to-end) if we think that code would break simulation tests - if so maybe we should not populate the version vector on the sequencer (or protect that code with the knob) and then accessing the version vector here won't have any effect.

Reply:
When I debug failed simulation tests, I like to have a single switch to turn off the feature. This way I can easily run with and then without the feature enabled. 
>if so maybe we should not populate the version vector on the sequencer 

yes, I'd only put the switch around code that is impactful, i.e. not everywhere.

Reply:
> Ok, will put the switch around code that populates the version vector on the sequencer.

This part has been done already (by Dan):

```
if (SERVER_KNOBS->ENABLE_VERSION_VECTOR && req.writtenTags.present()) {
					// NB: this if-condition is not needed after wait-for-prev is ported to this branch
					if (req.version > self->ssVersionVector.maxVersion) {
						// TraceEvent("Received ReportRawCommittedVersionRequest").detail("Version",req.version);
						self->ssVersionVector.setVersion(req.writtenTags.get(), req.version);
					}


```

Reply:
[others](https://github.com/apple/foundationdb/pull/5182)